File: javascript/example_code/s3/s3_setcors.js
/**
 * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * This file is licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License. A copy of
 * the License is located at
 *
 * http://aws.amazon.com/apache2.0/
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
*/

//snippet-sourcedescription:[s3_setcors.js demonstrates how to set the CORS configuration of an Amazon S3 bucket]
//snippet-service:[s3]
//snippet-keyword:[JavaScript]
//snippet-keyword:[Code Sample]
//snippet-keyword:[Amazon S3]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[2018-06-02]
//snippet-sourceauthor:[AWS-JSDG]

// ABOUT THIS NODE.JS SAMPLE: This sample is part of the SDK for JavaScript Developer Guide topic at
// https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/s3-example-configuring-buckets.html

// snippet-start:[s3.JavaScript.cors.putBucketCors]
// Load the AWS SDK for Node.js
var AWS = require('aws-sdk');
// Set the region 
AWS.config.update({region: 'REGION'});

// Create S3 service object
s3 = new AWS.S3({apiVersion: '2006-03-01'});

// Create initial parameters JSON for putBucketCors
var thisConfig = {
  AllowedHeaders:["Authorization"],
  AllowedMethods:[],
  AllowedOrigins:["*"],
  ExposeHeaders:[],
  MaxAgeSeconds:3000
};

// Assemble the list of allowed methods based on command line parameters
var allowedMethods = [];
process.argv.forEach(function (val, index, array) {
  if (val.toUpperCase() === "POST") {allowedMethods.push("POST")};
  if (val.toUpperCase() === "GET") {allowedMethods.push("GET")};
  if (val.toUpperCase() === "PUT") {allowedMethods.push("PUT")};
  if (val.toUpperCase() === "PATCH") {allowedMethods.push("PATCH")};
  if (val.toUpperCase() === "DELETE") {allowedMethods.push("DELETE")};
  if (val.toUpperCase() === "HEAD") {allowedMethods.push("HEAD")};
});

// Copy the array of allowed methods into the config object
thisConfig.AllowedMethods = allowedMethods;
// Create array of configs then add the config object to it
var corsRules = new Array(thisConfig);

// Create CORS params
var corsParams = {Bucket: process.argv[2], CORSConfiguration: {CORSRules: corsRules}};

// set the new CORS configuration on the selected bucket
s3.putBucketCors(corsParams, function(err, data) {
  if (err) {
    // display error message
    console.log("Error", err);
  } else {
    // update the displayed CORS config for the selected bucket
    console.log("Success", data);
  }
});
// snippet-end:[s3.JavaScript.cors.putBucketCors]

File: go/example_code/s3/s3_set_cors.go
// snippet-comment:[These are tags for the AWS doc team's sample catalog. Do not remove.]
// snippet-sourceauthor:[Doug-AWS]
// snippet-sourcedescription:[Sets CORS permission on an S3 bucket.]
// snippet-keyword:[Amazon Simple Storage Service]
// snippet-keyword:[Amazon S3]
// snippet-keyword:[PutBucketCors function]
// snippet-keyword:[Go]
// snippet-sourcesyntax:[go]
// snippet-service:[s3]
// snippet-keyword:[Code Sample]
// snippet-sourcetype:[full-example]
// snippet-sourcedate:[2020-1-6]
/*
   Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.

   This file is licensed under the Apache License, Version 2.0 (the "License").
   You may not use this file except in compliance with the License. A copy of
   the License is located at

    http://aws.amazon.com/apache2.0/

   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied. See the License for the
   specific language governing permissions and limitations under the License.
*/
// snippet-start:[s3.go.set_cors.complete]
package main

// snippet-start:[s3.go.set_cors.imports]
import (
    "flag"
    "fmt"
    "os"
    "strings"

    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/s3"
)
// snippet-end:[s3.go.set_cors.imports]

// Configures CORS rules for a bucket by setting the allowed
// HTTP methods.
//
// Requires the bucket name, and can also take a space separated
// list of HTTP methods.
//
// Usage:
//    go run s3_set_cors.go -b BUCKET_NAME get put
func main() {
    // snippet-start:[s3.go.set_cors.vars]
    bucketPtr := flag.String("b", "", "Bucket to set CORS on, (required)")

    flag.Parse()

    if *bucketPtr == "" {
        exitErrorf("-b <bucket> Bucket name required")
    }

    methods := filterMethods(flag.Args())
    // snippet-end:[s3.go.set_cors.vars]

    // Initialize a session
    // snippet-start:[s3.go.set_cors.session]
    sess := session.Must(session.NewSessionWithOptions(session.Options{
        SharedConfigState: session.SharedConfigEnable,
    }))

    svc := s3.New(sess)
    // snippet-end:[s3.go.set_cors.session]

    // Create a CORS rule for the bucket
    // snippet-start:[s3.go.set_cors.rule]
    rule := s3.CORSRule{
        AllowedHeaders: aws.StringSlice([]string{"Authorization"}),
        AllowedOrigins: aws.StringSlice([]string{"*"}),
        MaxAgeSeconds:  aws.Int64(3000),

        // Add HTTP methods CORS request that were specified in the CLI.
        AllowedMethods: aws.StringSlice(methods),
    }
    // snippet-end:[s3.go.set_cors.rule]

    // Create the parameters for the PutBucketCors API call, add add
    // the rule created to it.
    // snippet-start:[s3.go.set_cors.put]
    params := s3.PutBucketCorsInput{
        Bucket: bucketPtr,
        CORSConfiguration: &s3.CORSConfiguration{
            CORSRules: []*s3.CORSRule{&rule},
        },
    }

    _, err := svc.PutBucketCors(&params)
    if err != nil {
        // Print the error message
        exitErrorf("Unable to set Bucket %q's CORS, %v", *bucketPtr, err)
    }

    // Print the updated CORS config for the bucket
    fmt.Printf("Updated bucket %q CORS for %v\n", *bucketPtr, methods)
    // snippet-end:[s3.go.set_cors.put]
}

// snippet-start:[s3.go.set_cors.exit]
func exitErrorf(msg string, args ...interface{}) {
    fmt.Fprintf(os.Stderr, msg+"\n", args...)
    os.Exit(1)
}
// snippet-end:[s3.go.set_cors.exit]

// snippet-start:[s3.go.set_cors.filter]
func filterMethods(methods []string) []string {
    filtered := make([]string, 0, len(methods))
    for _, m := range methods {
        v := strings.ToUpper(m)
        switch v {
        case "POST", "GET", "PUT", "PATCH", "DELETE":
            filtered = append(filtered, v)
        }
    }

    return filtered
}
// snippet-end:[s3.go.set_cors.filter]
// snippet-end:[s3.go.set_cors.complete]
